File: snapir/topology.py
"""The connections the surveyor actually drew.

The plain room CSV is a bag of points with no connectivity. The `_FUKOKU.csv`
beside it carries the lines: a `Line start` / `Line end` pair per row. That file
holds the whole topology of the room, already closed:

    P_001 -> P_002 -> ... -> P_011 -> P_001      the floor ring
    P_012 -> P_013 -> ... -> P_022 -> P_012      the ceiling ring
    P_023 -> P_024 -> P_025 -> P_026 -> P_023    a door, as a closed loop
    P_013 -> P_001, P_014 -> P_002, ...          floor to ceiling links

Reading it means the room is described rather than guessed. Nothing here
invents a connection; every line drawn in the app comes from this file or from
the operator.
"""
from __future__ import annotations

import csv
from dataclasses import dataclass, field
from pathlib import Path

# Two points this close in Z belong to the same level, so the line between
# them lies flat rather than rising.
LEVEL_DZ = 35.0


@dataclass
class Topology:
    segments: list[tuple[str, str]] = field(default_factory=list)
    floor_ring: list[str] = field(default_factory=list)
    ceiling_ring: list[str] = field(default_factory=list)
    openings: list[list[str]] = field(default_factory=list)
    links: list[tuple[str, str]] = field(default_factory=list)   # floor to ceiling

    @property
    def found(self) -> bool:
        return len(self.segments) > 0


def fukoku_path(room_csv: str | Path) -> Path:
    p = Path(room_csv)
    return p.with_name(f"{p.stem}_FUKOKU.csv")


def read_segments(room_csv: str | Path) -> list[tuple[str, str]]:
    """Every line the operator drew, as point-name pairs."""
    path = fukoku_path(room_csv)
    if not path.exists():
        return []

    out: list[tuple[str, str]] = []
    with path.open(encoding="utf-8-sig", newline="") as fh:
        for row in csv.reader(fh, delimiter=";"):
            if len(row) < 2:
                continue
            a, b = row[0].strip(), row[1].strip()
            if not a or not b or a == "Line start":
                continue          # a standalone point, or the header
            out.append((a, b))
    return out


def _adjacency(segments, keep=None) -> dict[str, set[str]]:
    adj: dict[str, set[str]] = {}
    for a, b in segments:
        if keep and not keep(a, b):
            continue
        adj.setdefault(a, set()).add(b)
        adj.setdefault(b, set()).add(a)
    return adj


def _components(adj: dict[str, set[str]]) -> list[set[str]]:
    seen: set[str] = set()
    out = []
    for start in adj:
        if start in seen:
            continue
        stack, comp = [start], set()
        while stack:
            n = stack.pop()
            if n in comp:
                continue
            comp.add(n)
            stack.extend(adj[n] - comp)
        seen |= comp
        out.append(comp)
    return out


def _walk_cycle(adj: dict[str, set[str]], comp: set[str]) -> list[str]:
    """Order a component that is a simple closed loop. Empty if it is not."""
    if len(comp) < 3 or any(len(adj[n]) != 2 for n in comp):
        return []
    start = min(comp)
    ring = [start]
    prev, cur = None, start
    while True:
        nxt = next((n for n in adj[cur] if n != prev), None)
        if nxt is None:
            return []
        if nxt == start:
            return ring if len(ring) == len(comp) else []
        ring.append(nxt)
        prev, cur = cur, nxt
        if len(ring) > len(comp):
            return []


def build(segments, points: dict[str, tuple[float, float, float]]) -> Topology:
    """Sort the drawn lines into rings, openings and vertical links."""
    topo = Topology(segments=list(segments))
    known = [s for s in segments if s[0] in points and s[1] in points]
    if not known:
        return topo

    z = {n: points[n][2] for n in points}

    def level(a: str, b: str) -> bool:
        return abs(z[a] - z[b]) <= LEVEL_DZ

    # Flat lines first. Rings live entirely at one level.
    flat = _adjacency(known, keep=level)
    rings: list[list[str]] = []
    for comp in _components(flat):
        ring = _walk_cycle(flat, comp)
        if len(ring) >= 3:
            rings.append(ring)

    if rings:
        by_height = sorted(rings, key=lambda r: sum(z[n] for n in r) / len(r))
        topo.floor_ring = by_height[0]
        if len(by_height) > 1:
            top = by_height[-1]
            # A ring well above the floor one, with a matching corner count,
            # is the ceiling. Anything else is left alone.
            lo = sum(z[n] for n in topo.floor_ring) / len(topo.floor_ring)
            hi = sum(z[n] for n in top) / len(top)
            if hi - lo > 120.0:
                topo.ceiling_ring = top

    ring_nodes = set(topo.floor_ring) | set(topo.ceiling_ring)

    # Whatever is left over and forms its own closed loop is an opening: two
    # jambs and the sill and head lines between them.
    full = _adjacency(known)
    for comp in _components(full):
        if comp & ring_nodes:
            continue
        loop = _walk_cycle(full, comp)
        if len(loop) >= 4:
            topo.openings.append(loop)

    # Risers that join a floor corner to a ceiling corner. These are the links
    # the operator drew by hand, and the only ones the app will show.
    floor_set, ceil_set = set(topo.floor_ring), set(topo.ceiling_ring)
    for a, b in known:
        if level(a, b):
            continue
        if a in floor_set and b in ceil_set:
            topo.links.append((a, b))
        elif b in floor_set and a in ceil_set:
            topo.links.append((b, a))

    return topo

File: native/src/topology.cpp
#include "snapir/topology.hpp"
#include "snapir/csv.hpp"
#include <algorithm>
#include <cmath>
#include <set>
#include <unordered_map>

namespace snapir {
namespace {

// Adjacency that remembers the order names were first seen, so component and
// opening ordering is reproducible rather than hash-dependent.
struct Adjacency {
  std::vector<std::string> order;
  std::unordered_map<std::string, std::set<std::string>> map;

  void add(const std::string& a, const std::string& b) {
    if (map.find(a) == map.end()) order.push_back(a);
    map[a].insert(b);
    if (map.find(b) == map.end()) order.push_back(b);
    map[b].insert(a);
  }
  const std::set<std::string>& at(const std::string& n) const { return map.at(n); }
};

std::vector<std::set<std::string>> components(const Adjacency& adj) {
  std::set<std::string> seen;
  std::vector<std::set<std::string>> out;
  for (const auto& start : adj.order) {
    if (seen.count(start)) continue;
    std::vector<std::string> stack{start};
    std::set<std::string> comp;
    while (!stack.empty()) {
      const std::string n = stack.back();
      stack.pop_back();
      if (comp.count(n)) continue;
      comp.insert(n);
      for (const auto& m : adj.at(n))
        if (!comp.count(m)) stack.push_back(m);
    }
    seen.insert(comp.begin(), comp.end());
    out.push_back(std::move(comp));
  }
  return out;
}

// Order a component that is a simple closed loop. Empty if it is not.
std::vector<std::string> walk_cycle(const Adjacency& adj,
                                    const std::set<std::string>& comp) {
  if (comp.size() < 3) return {};
  for (const auto& n : comp)
    if (adj.at(n).size() != 2) return {};

  const std::string start = *comp.begin();  // std::set is sorted: min(comp)
  std::vector<std::string> ring{start};
  std::string prev, cur = start;
  bool has_prev = false;
  while (true) {
    std::string nxt;
    bool found = false;
    for (const auto& n : adj.at(cur)) {
      if (has_prev && n == prev) continue;
      nxt = n;
      found = true;
      break;
    }
    if (!found) return {};
    if (nxt == start) return ring.size() == comp.size() ? ring : std::vector<std::string>{};
    ring.push_back(nxt);
    prev = cur;
    has_prev = true;
    cur = nxt;
    if (ring.size() > comp.size()) return {};
  }
}

double mean_z(const std::vector<std::string>& ring,
              const std::unordered_map<std::string, double>& z) {
  double s = 0.0;
  for (const auto& n : ring) s += z.at(n);
  return s / static_cast<double>(ring.size());
}

}  // namespace

std::filesystem::path fukoku_path(const std::string& room_csv) {
  // The room path arrives as UTF-8; u8path keeps the Turkish room names intact
  // rather than losing them through the console codepage.
  const std::filesystem::path p = std::filesystem::u8path(room_csv);
  // operator/ would read a narrow string back as the ANSI codepage, which
  // loses the Turkish room names, so the joined name goes through u8path too.
  return p.parent_path() / std::filesystem::u8path(p.stem().u8string() + "_FUKOKU.csv");
}

std::vector<Segment> read_segments(const std::string& room_csv) {
  const std::filesystem::path path = fukoku_path(room_csv);
  std::vector<Segment> out;
  if (!file_exists(path)) return out;

  for (const auto& row : read_csv(path, ';')) {
    if (row.size() < 2) continue;
    const std::string a = trim(row[0]), b = trim(row[1]);
    if (a.empty() || b.empty() || a == "Line start") continue;  // point, or header
    out.emplace_back(a, b);
  }
  return out;
}

Topology build_topology(const std::vector<Segment>& segments,
                        const std::map<std::string, std::array<double, 3>>& points) {
  Topology topo;
  topo.segments = segments;

  std::vector<Segment> known;
  for (const auto& s : segments)
    if (points.count(s.first) && points.count(s.second)) known.push_back(s);
  if (known.empty()) return topo;

  std::unordered_map<std::string, double> z;
  for (const auto& kv : points) z[kv.first] = kv.second[2];

  const auto level = [&z](const std::string& a, const std::string& b) {
    return std::abs(z.at(a) - z.at(b)) <= kLevelDz;
  };

  // Flat lines first. Rings live entirely at one level.
  Adjacency flat;
  for (const auto& s : known)
    if (level(s.first, s.second)) flat.add(s.first, s.second);

  std::vector<std::vector<std::string>> rings;
  for (const auto& comp : components(flat)) {
    const auto ring = walk_cycle(flat, comp);
    if (ring.size() >= 3) rings.push_back(ring);
  }

  if (!rings.empty()) {
    std::stable_sort(rings.begin(), rings.end(),
                     [&z](const std::vector<std::string>& a,
                          const std::vector<std::string>& b) {
                       return mean_z(a, z) < mean_z(b, z);
                     });
    topo.floor_ring = rings.front();
    if (rings.size() > 1) {
      // A ring well above the floor one, with a matching corner count, is the
      // ceiling. Anything else is left alone.
      const double lo = mean_z(topo.floor_ring, z);
      const double hi = mean_z(rings.back(), z);
      if (hi - lo > 120.0) topo.ceiling_ring = rings.back();
    }
  }

  std::set<std::string> ring_nodes(topo.floor_ring.begin(), topo.floor_ring.end());
  ring_nodes.insert(topo.ceiling_ring.begin(), topo.ceiling_ring.end());

  // Whatever is left over and forms its own closed loop is an opening: two
  // jambs and the sill and head lines between them.
  Adjacency full;
  for (const auto& s : known) full.add(s.first, s.second);
  for (const auto& comp : components(full)) {
    bool touches_ring = false;
    for (const auto& n : comp)
      if (ring_nodes.count(n)) { touches_ring = true; break; }
    if (touches_ring) continue;
    const auto loop = walk_cycle(full, comp);
    if (loop.size() >= 4) topo.openings.push_back(loop);
  }

  // Risers that join a floor corner to a ceiling corner. These are the links
  // the operator drew by hand, and the only ones the app will show.
  const std::set<std::string> floor_set(topo.floor_ring.begin(), topo.floor_ring.end());
  const std::set<std::string> ceil_set(topo.ceiling_ring.begin(), topo.ceiling_ring.end());
  for (const auto& s : known) {
    if (level(s.first, s.second)) continue;
    if (floor_set.count(s.first) && ceil_set.count(s.second))
      topo.links.emplace_back(s.first, s.second);
    else if (floor_set.count(s.second) && ceil_set.count(s.first))
      topo.links.emplace_back(s.second, s.first);
  }

  return topo;
}

}  // namespace snapir
